hal/reset_reason_api.h

```diff
@@ -0,0 +1,67 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_RESET_REASON_API_H
+#define MBED_RESET_REASON_API_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum {
+  RESET_REASON_POWER_ON  = (1 << 0), /**< Set when power is initially applied to the board. The power-on-reset circuit causes a POWER_ON reset when this occurs */
+  RESET_REASON_PIN_RESET = (1 << 1), /**< Set when a reset is triggered by the hardware pin on the board */
+  RESET_REASON_BROWN_OUT = (1 << 2), /**< Triggered when the voltage drops below the low voltage detect (LVD) threshold the system will be held in a reset until the voltage rises above the threshold */
+  RESET_REASON_SOFTWARE  = (1 << 3), /**< Set during software reset, typically triggered by writing the SYSRESETREQ bit in the Application Interrupt and Reset Control register */
+  RESET_REASON_WATCHDOG  = (1 << 4), /**< Set when a running watchdog timer fails to be refreshed */
+  RESET_REASON_LOCKUP    = (1 << 5), /**< Set when the core is locked because of an unrecoverable exception */
+  RESET_REASON_MULTIPLE  = (1 << 6), /**< Set if multiple reset reasons are set within the board. Occurs when the reset reason registers aren't cleared between resets */
+  RESET_REASON_PLATFORM  = (1 << 7), /**< Platform specific reset reason not captured in this enum */
+  RESET_REASON_UNKNOWN   = (1 << 8)  /**< Unknown or unreadable reset reason **/
```

Comment:
Are we assuming platforms can add their own reasons and OR `RESET_REASON_PLATFORM` and some reason higher than UNKNOWN? If so we should make sure enum is big enough by adding some `LAST` and document that.

Reply:
I had initially intended it be used instead of allowing platforms to add their own reasons. But, it doesn't seem like there are going to be too many more exotic platform specific ones (K64F adds about 7), so there shouldn't be any harm removing the PLATFORM field and adding platform specific ones.

Reply:
We are trying to support all the boards equally, I don't want features that are on one board only. I would rather enable platform/apps to go into platform details if they really want to (and don't need to be portable) by having the PLATFORM and space for board specific cases.

Reply:
> I had initially intended it be used instead of allowing platforms to add their own reasons. 

I believe this was always what we do not wanted achieve - to have non portable programs (if a target provides more reasons, use their own SDK  to get them, HAL does not support platform specific ). If in the user space is a switch to get a reason and its platform specific, we would end up with:

```
switch reason:

#ifdef this target
    case this_target_only_reason:
        //handle it and break
#endif
```

Similar issue is in the pullmodes, see the report https://github.com/ARMmbed/mbed-os/issues/1064.

Another question is  `RESET_REASON_MULTIPLE  ` is needed? cant this be removed and handle via get/clear (we can not end up with multiple reasons, not sure just multiple does not seem OK here) ? If multiple, what value for an application does this have?

Reply:
Good point, let us keep the platform, but lets be explicit we son't do anything more about it and at this point behavior is not defined.

Reply:
> I believe this was always what we do not wanted achieve - to have non portable programs (if a target provides more reasons, use their own SDK to get them, HAL does not support platform specific). If in the user space is a switch to get a reason and its platform specific, we would end up in an application:

Not wanting to support platform specific things in the HAL is reasonable. That was the intention of the PLATFORM return value. The application can use the platform specific SDK to get the platform specific reset reason after receiving it.

> Another question is RESET_REASON_MULTIPLE is needed? cant this be removed and handle via get/clear (we can not end up with multiple reasons, not sure just multiple does not seem OK here) ? If multiple, what value for an application does this have?

Currently, the reset reason is not implicitly cleared when get is called, so it is possible for the reset registers to have multiple reset reasons, I'm not sure if it's necessary to have a specific enum value for that, it just distinguishes it from an UNKNOWN.

Reply:
Also if platform itself is not clearing the reason over restarts, and the app is not calling `get` after every reset, we may end up in a situation with multiple reasons in the register.